### DISCUSSION TITLE
How to add self loops for nodes without self-edges?

### DISCUSSION THREAD

**Discussion by wgeul** [pyg-team/pytorch_geometric#7904] [Q&A] [answered]
I'm using `HeteroData` objects and I'm experimenting with self loops, adding the self loops like this:

`data = T.AddSelfLoops()(data)`

I notice the `edge_index` size expands for existing self-linked edges `('source', 'to', 'source')`, but I don't see new edges being created for self-linked edges that were not yet in the data object, e.g. `('target', 'to', 'target')`:

```
HeteroData(
source= { x=[..]},
target = { x =[..]},
('source', 'to', 'source') = { edge_index = [2, ..]}
)
```

So no addition of `('target', 'to', 'target')`

What's the best way to add self loops for `target`?

**Comment by wgeul** [comment]
I presume this would do, please confirm:
```
data = T.AddSelfLoops()(data)
for node in data.metadata()[0]:
                if (node, "to", node) not in data.metadata()[1]:
                    dim = data[node].x.shape[0]
                    data[(node, "to", node)].edge_index = stack(
                        (arange(dim), arange(dim))
                    )
```

To be honest I expected to get self loops for all node types as default behavior.

**Reply by wsad1** [reply, marked as answer]
The edge type `('target', 'to', 'target')` does not exist, so `AddSelfLoops` will no add the edge type and add self loops to it. 
Your approach above looks good. 
You could also try
```
data[('target', 'to', 'target')].edge_index = torch.empty(2,0)
data = T.AddSelfLoops()(data)
```

**Reply by wgeul** [reply]
Thanks @wsad1
